src/flp/types_utilities.rs

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Some functions shared by several [`Type`](crate::flp::Type) implementations.
+
+use crate::field::FieldElement;
+use crate::flp::{FlpError, Gadget, Type};
+use std::convert::TryFrom;
+
+/// Interpret `bits` as `F::Integer` if it's representable in that type and smaller than the field modulus.
+pub(crate) fn valid_int_from_usize<F: FieldElement>(bits: usize) -> Result<F::Integer, FlpError> {
+    let bits_int = F::Integer::try_from(bits).map_err(|err| {
+        FlpError::Encode(format!(
+            "bit length ({}) cannot be represented as a field element: {:?}",
+            bits, err,
+        ))
+    })?;
+
+    if F::modulus() >> bits_int == F::Integer::from(F::zero()) {
+        return Err(FlpError::Encode(format!(
+            "bit length ({}) exceeds field modulus",
+            bits,
+        )));
+    }
+    Ok(bits_int)
+}
+
+/// Encode `input` as `bits`-bit vector of elements of `T::Field` if it's small enough
+/// to be represented with that many bits.
+///
+/// # Arguments
+///
+/// * `input` - The field element to encode
+/// * `bits` - The number of bits to use for the encoding
+/// * `max_input` - The maximal number representable with `bits` bits. Equal to `bits^2-1`
+pub(crate) fn encode_into_bitvector_representation<F: FieldElement>(
```

Comment:
This function doesn't really operate on any `Type`, so I'd say maybe it should be put into `field.rs`. I also think you could refactor it into an associated function on trait `FieldElement` by inserting the following function definition into `pub trait FieldElement`:
```
    /// Encode `input` as `bits`-bit vector of elements of `T::Field` if it's small enough
    /// to be represented with that many bits.
    ///
    /// # Arguments
    ///
    /// * `input` - The field element to encode
    /// * `bits` - The number of bits to use for the encoding
    /// * `max_input` - The maximal number representable with `bits` bits. Equal to `bits^2-1`
    fn encode_into_bitvector_representation(
        input: &Self::Integer,
        bits: usize,
        max_input: Self::Integer,
    ) -> Result<Vec<Self>, FieldError> {
        if *input > max_input {
            return Err(FieldError::InputSizeMismatch);
        }

        let one = Self::Integer::from(Self::one());
        let mut encoded = Vec::with_capacity(bits);
        for l in 0..bits {
            let l = Self::Integer::try_from(l).unwrap();
            let w = Self::from((*input >> l) & one);
            encoded.push(w);
        }
        Ok(encoded)
    }
```

I think this makes it easier for users of the library to discover these functions, since they'll now show up when they read the documentation on [trait `FieldElement`](https://docs.rs/prio/latest/prio/field/trait.FieldElement.html), alongside the other default trait functions provided there.

I suggest the same approach for `valid_int_from_usize`, `decode_from_bitvector_representation`, `call_gadget_on_vec_entries`, `decode_result` and `decode_result_vec`.